keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1182 @@
+applications, but for HA applications as well. 
+
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin I want to have a simple interface to initiate a node drain/maintenance without
+any required manual interventions. I also want to observe the node drain via the API and check on
+its progress. I also want to be able to discover workloads that are blocking the node drain.
+
+#### Story 2
+
+As an application owner, I want to run single replica applications without disruptions and have the
+ability to easily migrate the workload pods from one node to another.
+
+#### Story 3
+
+Cluster or node autoscalers that take on the role of `kubectl drain` want to signal the intent to
+drain a node using the same API and provide a similar experience to the CLI counterpart.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Kubectl
+
+`kubectl drain` command will be changed to create a NodeMaintenance object after marking node
+unschedulable. We will also change the implementation to skip applications that support
+workload migration. At first, we will just check deployments with a positive
+`.spec.strategy.rollingUpdate.maxSurge` value. To make this extensible, we propose a new Surge API
+that could be checked before disrupting any workload (see below). If the cluster,
+doesn't support the NodeMaintenance API, kubectl will perform the node drain in a backwards
+compatible way.
+
+`kubectl uncordon` will be amended with `--node-maintenance` option. This can be used to delete the
+NodeMaintenance object together with marking the node as schedulable again. The default value of
+this flag will be `false`.
+
+### NodeMaintenance API
+
+NodeMaintenance objects serve as an intent to remove or migrate pods from a particular node. We will
+also support setting a reason for the maintenance.
+
+```golang
+
+type NodeMaintenance struct {
+    ...
+    Spec NodeMaintenanceSpec
+    Status NodeMaintenanceStatus
+}
+
+type NodeMaintenanceSpec struct {
+    NodeName string
+    Reason string
+}
+
+type NodeMaintenanceStatus struct {
+    // number of pods this node maintenance is requesting to terminate
+    TargetedPods int32
+}
+```
+
+The NodeMaintenance API is being introduced with a minimal set of features, but it is expected that
+additional features will be added to the API in the future.
```

Comment:
Not needed for this PR; I'll mention it anyway in case it's useful.

Sometimes people want node maintenance to trigger a drain; other times, just a cordon.

Reply:
You mean cordon in a sense that makes the node unschedulable? I am not sure if this should be part of the API if users can just use `kubectl cordon` or edit nodes manually.

Reply:
**Future use**

One NodeMaintenance with a selector could cordon all nodes with some property (eg: all the AArch64 nodes that also have `feature.node.kubernetes.io/example-thing` labels).

Reply:
I have updated the KEP to include the phases of node maintenance. One of them is cordon. These phases can now be toggled with the NodeMaintenance API.